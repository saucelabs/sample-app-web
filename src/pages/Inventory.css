.product_label {
  position: absolute;
  left: 110px;
  top: 25px;
  font-size: 32px;
  font-weight: 100;
  color: #fff;
}

.inventory_container {
  overflow: hidden;
  margin-bottom: 30px;
  padding-top: 40px;
}

.inventory_list {
  max-width: 1280px;
  margin: 0 auto;
  display: flex;
  justify-content: space-between;
  flex-flow: row wrap;
  padding: 0 15px 30px;
}
