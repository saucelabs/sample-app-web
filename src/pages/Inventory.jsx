import React, { useState } from "react";
import { withRouter } from "react-router-dom";
import { isPerformanceGlitchUser, isProblemUser } from "../utils/Credentials";
import { InventoryData } from "../utils/InventoryData.js";
import InventoryListItem from "../components/InventoryListItem";
import SwagLabsFooter from "../components/Footer";
import HeaderContainer from "../components/HeaderContainer";
import { sortAsc, sortDesc, sortHiLo, sortLoHi } from "../utils/Sorting";
import Select from "../components/Select";
import "./Inventory.css";

const Inventory = () => {
  const [inventoryList, setInventoryList] = useState(
    sortAsc(InventoryData, "name")
  );
  const [activeOption, setActiveOption] = useState("az");
  /* istanbul ignore next */
  const startPerformanceGlitch = (duration) => {
    const start = new Date().getTime();
    while (new Date().getTime() < start + duration) {
      // PageLoad increases
    }
  };

  /* istanbul ignore next */
  if (isPerformanceGlitchUser()) {
    startPerformanceGlitch(5000);
  }

  /**
   * @TODO:
   * This can't be tested yet because enzyme currently doesn't support ReactJS17,
   * see https://github.com/enzymejs/enzyme/issues/2429.
   * This means we can't fully mount the component and test all rendered components
   * and functions
   */
  /* istanbul ignore next */
  const sortByOption = (event) => {
    if (isProblemUser()) {
      // Bail out now if we're problem user so that we have a behaviour which is broken in Chrome only for sort.
      // select option onclick is not supported in Chrome but works in IE and FF
      return;
    }

    setActiveOption(event.target.value);

    switch (event.target.value) {
      case "az":
        setInventoryList(sortAsc(InventoryData, "name"));
        break;
      case "za":
        setInventoryList(sortDesc(InventoryData, "name"));
        break;
      case "hilo":
        setInventoryList(sortHiLo(InventoryData, "price"));
        break;
      case "lohi":
        setInventoryList(sortLoHi(InventoryData, "price"));
        break;
      default:
        return;
    }
  };

  return (
    <div id="page_wrapper" className="page_wrapper">
      <div id="contents_wrapper">
        <HeaderContainer
          secondaryTitle="Products"
          secondaryHeaderBot
          secondaryRightComponent={
            <Select
              activeOption={activeOption}
              options={[
                { key: "az", value: "Name (A to Z)" },
                { key: "za", value: "Name (Z to A)" },
                { key: "lohi", value: "Price (low to high)" },
                { key: "hilo", value: "Price (high to low)" },
              ]}
              onChange={sortByOption}
              testId="product_sort_container"
            />
          }
        />
        <div id="inventory_container">
          <div>
            <div id="inventory_container" className="inventory_container">
              <div className="inventory_list">
                {inventoryList.map((item, i) => {
                  return (
                    <InventoryListItem
                      key={item.id}
                      id={item.id}
                      image_url={item.image_url}
                      name={item.name}
                      desc={item.desc}
                      price={item.price}
                    />
                  );
                })}
              </div>
            </div>
          </div>
        </div>
      </div>
      <SwagLabsFooter />
    </div>
  );
};

export default withRouter(Inventory);
