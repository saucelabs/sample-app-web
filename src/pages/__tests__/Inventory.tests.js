import React, { useState as useStateMock } from "react";
import { shallow } from "enzyme";
import Inventory from "../Inventory";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("Inventory", () => {
  const setState = jest.fn();

  beforeEach(() => {
    useStateMock.mockImplementation((init) => [init, setState]);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const wrapper = shallow(<Inventory.WrappedComponent />);
    expect(wrapper).toMatchSnapshot();
  });
});
