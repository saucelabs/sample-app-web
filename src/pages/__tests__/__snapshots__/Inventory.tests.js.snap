// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inventory should render correctly 1`] = `
<div
  className="page_wrapper"
  id="page_wrapper"
>
  <div
    id="contents_wrapper"
  >
    <HeaderContainer
      secondaryHeaderBot={true}
      secondaryRightComponent={
        <Select
          activeOption="az"
          onChange={[Function]}
          options={
            Array [
              Object {
                "key": "az",
                "value": "Name (A to Z)",
              },
              Object {
                "key": "za",
                "value": "Name (Z to A)",
              },
              Object {
                "key": "lohi",
                "value": "Price (low to high)",
              },
              Object {
                "key": "hilo",
                "value": "Price (high to low)",
              },
            ]
          }
          testId="product_sort_container"
        />
      }
      secondaryTitle="Products"
    />
    <div
      id="inventory_container"
    >
      <div>
        <div
          className="inventory_container"
          id="inventory_container"
        >
          <div
            className="inventory_list"
          >
            <withRouter(InventoryListItem)
              desc="carry.allTheThings() with the sleek, streamlined Sly Pack that melds uncompromising style with unequaled laptop and tablet protection."
              id={4}
              image_url="sauce-backpack-1200x1500.jpg"
              key="4"
              name="Sauce Labs Backpack"
              price={29.99}
            />
            <withRouter(InventoryListItem)
              desc="A red light isn't the desired state in testing but it sure helps when riding your bike at night. Water-resistant with 3 lighting modes, 1 AAA battery included."
              id={0}
              image_url="bike-light-1200x1500.jpg"
              key="0"
              name="Sauce Labs Bike Light"
              price={9.99}
            />
            <withRouter(InventoryListItem)
              desc="Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt."
              id={1}
              image_url="bolt-shirt-1200x1500.jpg"
              key="1"
              name="Sauce Labs Bolt T-Shirt"
              price={15.99}
            />
            <withRouter(InventoryListItem)
              desc="It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office."
              id={5}
              image_url="sauce-pullover-1200x1500.jpg"
              key="5"
              name="Sauce Labs Fleece Jacket"
              price={49.99}
            />
            <withRouter(InventoryListItem)
              desc="Rib snap infant onesie for the junior automation engineer in development. Reinforced 3-snap bottom closure, two-needle hemmed sleeved and bottom won't unravel."
              id={2}
              image_url="red-onesie-1200x1500.jpg"
              key="2"
              name="Sauce Labs Onesie"
              price={7.99}
            />
            <withRouter(InventoryListItem)
              desc="This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton."
              id={3}
              image_url="red-tatt-1200x1500.jpg"
              key="3"
              name="Test.allTheThings() T-Shirt (Red)"
              price={15.99}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <SwagLabsFooter />
</div>
`;
